src/posit/connect/context.py

```diff
@@ -23,25 +27,27 @@ def wrapper(instance: ContextManager, *args, **kwargs):
     return decorator
 
 
-class Context(dict):
+class Context:
     def __init__(self, session: requests.Session, url: Url):
         self.session = session
         self.url = url
+        self._version: str | None
 
     @property
-    def version(self) -> Optional[str]:
-        try:
-            value = self["version"]
-        except KeyError:
-            endpoint = self.url + "server_settings"
-            response = self.session.get(endpoint)
-            result = response.json()
-            value = self["version"] = result.get("version")
+    def version(self) -> str | None:
+        if "_version" in self.__dict__:
+            return self._version
+
+        # Populate version
+        endpoint = self.url + "server_settings"
+        response = self.session.get(endpoint)
+        result = response.json()
+        value = self._version = result.get("version")
         return value
 
     @version.setter
     def version(self, value):
-        self["version"] = value
+        self._version = value
```

Comment:
How about...

```python
class Context:
    def __init__(self, session: requests.Session, url: Url):
        self.session = session
        self.url = url

    @property
    def version(self) -> str | None:
        if not hasattr(self, "_version"):
            endpoint = self.url + "server_settings"
            response = self.session.get(endpoint)
            result = response.json()
            self._version: str | None = result.get("version")

        return self._version

    @version.setter
    def version(self, value: str | None):
        self._version = value
```

`hasattr` is more common than checking the symbol table directly. And specifying the types on the setter and on L45 avoids the empty declaration in `__init__`